Allow ChatRequestParameters in AiServices chat method (#2410)

michal-bida opened the pull request:
## Issue
"Implements #2410"

## Change
Add option to set custom ChatRequestParameters when calling chat method on AiServices interface.


## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [X] There are no breaking changes
- [x] I have added unit and/or integration tests for my change
- [x] The tests cover both positive and negative cases
- [x] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [x] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [x] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
X - [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
X - [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

dliubarskyi commented:
Hi @michal-bida, thank you. But there are no tests and not clear how this should be configured.

michal-bida commented:
Yes, as I said in the feature request I believe this actual needs some design. 

I mean I can add the tests but I think adding these parameters one by one to the context is perhaps not the optimal solution? Unless it is and then I can surely finish this PR. But it would be good if someone more versed in langhain4j architecture would comment on it.

dliubarskyi commented:
@michal-bida sorry I did not see the issue (feature request) 🤦

michal-bida commented:
@dliubarskyi I have changed this PR completely - now an option to send ChatRequestParamters as part of AiServices.chat method is implemented. 
I have tried to add tests as well - however I am pretty sure this will need review and revision. :-)

leduckhc commented on docs/docs/tutorials/ai-services.md:
```diff
@@ -174,6 +174,46 @@ This annotation is necessary only when the `-parameters` option is *not* enabled
+
+    // Notice the second parameter is annotated with @ChatRequestParams 
+    // so you can pass custom parameters each time you call it.
+    String chat(@UserName String userMessage, @ChatRequestParams ChatRequestParameters params);
```
user name or user message?

michal-bida replied:
Good catch.

leduckhc commented on langchain4j-core/src/main/java/dev/langchain4j/model/chat/request/ChatRequestParams.java:
```diff
@@ -0,0 +1,11 @@
+package dev.langchain4j.model.chat.request;
```
Move under `dev.langchain4j.service` package

---

Btw, what's purpose of having (extra) interface ChatRequestParams? I understand purpose of having annotations for MemoryId, UserMessage, SystemMessage, UserName as those are Strings.

michal-bida replied:
Yeah... Good point - the annotation in this case is not needed - I will get rid of it and simply add support for passing ChatRequestParameters parameter...

leduckhc commented on langchain4j/src/main/java/dev/langchain4j/service/DefaultAiServices.java:
```diff
@@ -366,6 +383,20 @@ private Future<Moderation> triggerModerationIfNeeded(Method method, List<ChatMes
+        Parameter[] params = method.getParameters();
+        for (int i = 0; i < params.length; i++) {
+            if (params[i].isAnnotationPresent(ChatRequestParams.class)) {
+                // You could do extra checks (null-checking, type-checking) here if you want.
```
definitely null-checking, type-checking here. As I understand it ChatRequestParams is optional annotation, in which case wrap it in Optional<...>?

michal-bida replied:
I modified the implementation - ditched the annotation and just looking for the object class now - so this took care of this.

leduckhc reviewed (commented):
IMHO. There are many formatting changes. Please, shift those formatting changes into separate PR.

dliubarskyi commented on docs/docs/tutorials/ai-services.md:
```diff
@@ -174,6 +174,44 @@ This annotation is necessary only when the `-parameters` option is *not* enabled
+To achieve this, you can create a second parameter in your AI Service method that accepts ChatRequestParams—marked with a custom annotation (@ChatRequestParams). This tells LangChain4j to accept and merge these parameters during each invocation.
+
+:::note
+Note that toolsSpecification will be always overridden by the specification set on AiContext. 
```
hmm... as a user, I would probably expect `toolsSpecification` provided inside the `ChatRequestParameters` to override those configured on the AI Service

michal-bida replied:
Yeah, that makes total sense. I will change it!

michal-bida replied:
Changed in the code and updated documentation. PR Branch was updated.

dliubarskyi commented on langchain4j-core/src/main/java/dev/langchain4j/model/chat/request/ChatRequestParameters.java:
```diff
@@ -39,4 +38,18 @@ public interface ChatRequestParameters {
+     *           .build();
+     * </pre>
+     */
+    static DefaultChatRequestParameters.Builder<?> builder(ChatRequestParameters from) {
```
I plan to introduce [`ChatRequestParameters.overrideWith`](https://github.com/langchain4j/langchain4j/pull/2403/files#diff-d50ecd38db0ce63864d42b2d7b3b1ec05920bac29352162de2daaeb791be501dR67), perhaps it can be reused in this PR?

michal-bida replied:
Sure why not - whatever gets merged first. I can always get back to this and change it later.

michal-bida replied:
Rebased the MR using your override method - much cleaner now.

dliubarskyi commented:
@michal-bida thank you! I would not rush with this change and would like to brainstorm the API options.

cc @geoand WDYT?

geoand commented:
I actually like the idea a lot!

Another alternative would be to support some some kind of return type that would then allow customization and execution of the request, but I think this is simpler

Commit:
Allow ChatRequestParameters in AiServices chat method (#2410)

Add - option to pass custom ChatRequestParameters class to chat method
Add - test for the above
Change - changed validation error message and modified tests for it
Add - documentation

michal-bida commented:
@leduckhc @geoand  I have rebased the PR on latest greatest using newly introduced ChatRequestParameters override method (sorry for the delay, haven't noticed the drift). Everything looks fine - test wise.

What are the next steps now? Do you think it could be merged? I find being able to override default ChatRequestParams in AiServices.chat method quite useful...

geoand commented:
cc @mariofusco

dliubarskyi commented:
@michal-bida so sorry about the delay, this PR somehow got lost from my radar. Yes, this feature is for sure useful! If you are still interested, please fix the conflicts and ping me using @dliubarskyi. Thank you! 🙏

mariofusco commented:
Replaced by https://github.com/langchain4j/langchain4j/pull/4226
